### PR TITLE
Make tests consistent

### PR DESCRIPTION
**Pull request by JKutscha**
* don't rerun pit, if we run in a timeout
* make sure we are getting the menu from the workbench window

### CODE REVIEW

**Comment by JKutscha** [comment]
@LorenzoBettini or @echebbi would you have a look at these changes and merge please?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java [hunk 1]
```diff
@@ -39,7 +39,7 @@ public void renameClass(String newClassName) {
 
     private void clickRename() {
         SWTBotMenuHelper menuHelper = new SWTBotMenuHelper();
-        menuHelper.findMenu(bot.menu(REFACTOR), RENAME).click();
+        menuHelper.findMenu(menuHelper.findWorkbenchMenu(bot, REFACTOR), RENAME).click();
```
Comment (LorenzoBettini) [inline tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java:42]: I need to understand: why is that required? The situation where this is required looks like when there's another dialog open? Doesn't this represent a bug in the test?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java [hunk 2]
```diff
@@ -64,4 +81,17 @@ public SWTBotMenu findMenu(final SWTBotMenu parentMenu,
             return new SWTBotMenu(menuItem);
         }
     }
+
+    public SWTBotMenu findWorkbenchMenu(final SWTWorkbenchBot bot, final String menuString) {
+        for (SWTBotShell shell : bot.shells()) {
```
Comment (LorenzoBettini) [inline tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java:86]: If I understand correctly you want to make sure the Eclipse window is the one that's active right?
If so, SWTBot FAQ shows a way to do that, which is easier than this, and does not require to distinguish between Eclipse and Maven build (https://wiki.eclipse.org/SWTBot/Troubleshooting) something like this should force the focus on the Eclipse workbench:

```java
UIThreadRunnable.syncExec(new VoidResult() {
  public void run() {
    PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell().forceActive();
  }
});
```

**Review by LorenzoBettini — CHANGES_REQUESTED** [review]
I have a few doubts about the code: I think a few parts should be simplified as shown in the comments. I'd also like to know more about the situations they try to fix

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java [hunk 3]
```diff
@@ -33,16 +34,22 @@ public static void setupJavaProject() throws CoreException {
...
-        "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Foo |    7 | negated conditional");
-        pitclipseSteps.doubleClickMutationInMutationsView(mutation);
-        pitclipseSteps.mutationIsOpened(FOO_CLASS + ".java", 7);
-        mutation = fromMutationLine(
-        "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Bar |    7 | negated conditional");
-        pitclipseSteps.doubleClickMutationInMutationsView(mutation);
-        pitclipseSteps.mutationIsOpened(BAR_CLASS + ".java", 7);
+        final long timeOutBefore = SWTBotPreferences.TIMEOUT;
+        try {
+            SWTBotPreferences.TIMEOUT = 10000;
```
[the remark this answers is not in the document]
Comment (JKutscha) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java:39]: Because this timeout value is used for how long we are waiting for the pit run. Currently, the timeout is set to 5 Seconds and I turned it up to 10 seconds to give pit enough time to finish.

Should I use the search timeout instead, but this seems to be very long?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java (hunk 1 above)
Comment (JKutscha) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java:42]: This is technically not required in this particular case, but this implementation doesn't rely on the focus, which in some cases depends heavily on the window manager of the OS.

With this implementation, we just make sure to always use the correct shell to open the menu.

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java (hunk 2 above)
Comment (JKutscha) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java:86]: I don't want to use the focus as much as possible, because this can be a root of problems, as seen in #152.

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (LorenzoBettini) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java:39]: I think that we should use the same timeout consistently and make it configurable from the outside as we do right now... @echebbi what do you think?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java (hunk 1 above)
Comment (LorenzoBettini) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/behaviours/pageobjects/RefactorMenu.java:42]: Could you please describe the scenario where this is required? I seem to recall that you propose this PR because otherwise, a few tests in another PR were failing? Or did I not recall correctly?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java (hunk 2 above)
Comment (LorenzoBettini) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java:86]: OK, but if you need to access the workbench shell at least do that with the Eclipse API and not based on the name of the window (which is different in Eclipse and in Maven). Something similar to what I showed above. For example, this is taken from `org.eclipse.swtbot.eclipse.finder.DefaultWorkbench.getActiveWorkbenchWindowShell()` (Once you have a `Shell` you can construct an `SWTBotShell`):

```java
	private IWorkbenchWindow getActiveWorkbenchWindow() {
		return UIThreadRunnable.syncExec(bot.getDisplay(), new Result<IWorkbenchWindow>() {
			@Override
			public IWorkbenchWindow run() {
				return PlatformUI.getWorkbench().getActiveWorkbenchWindow();
			}
		});
	}

	private Widget getActiveWorkbenchWindowShell() {
		return getActiveWorkbenchWindow().getShell();
	}
```

you can see that the workbench window can be accessed programmatically independently from the window text.

Comment (LorenzoBettini) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java:86]: Please, also have a look at https://github.com/pitest/pitclipse/pull/152#issuecomment-896862926
from my experience when I get the test failure finding the menu is because SWTBot does not find the shell active (typically the Eclipse workbench) that's why I'm forcing the activation of the workbench.

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java [hunk 4]
```diff
@@ -37,11 +37,11 @@ public void selectMutationOpensTheClassAtTheRightLineNumber() throws CoreExcepti
         coverageReportGenerated(2, 80, 0, 6, 0);
         PitclipseSteps pitclipseSteps = new PitclipseSteps();
         PitMutation mutation = fromMutationLine(
-        "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Foo |    7 | negated conditional");
+                "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Foo |    7 | negated conditional");
         pitclipseSteps.doubleClickMutationInMutationsView(mutation);
         pitclipseSteps.mutationIsOpened(FOO_CLASS + ".java", 7);
         mutation = fromMutationLine(
-        "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Bar |    7 | negated conditional");
+                "SURVIVED    | " + TEST_PROJECT + " | foo.bar | foo.bar.Bar |    7 | negated conditional");
```
Comment (LorenzoBettini) [inline tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java:44]: are these changed indentations intentional?

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/PitResultNotifier.java [hunk 5]
```diff
@@ -138,7 +138,7 @@ public boolean test() {
                 public String getFailureMessage() {
                     return "No summary was generated after the specified time.";
                 }
-            }, (timeOut > 0) ? timeOut : SWTBotPreferences.TIMEOUT);
+            }, (timeOut > 0) ? timeOut : Long.parseLong(System.getProperty(SWTBotPreferences.KEY_TIMEOUT)));
```
Comment (LorenzoBettini) [inline tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/PitResultNotifier.java:141]: In SWTBot you have

```java
public static long TIMEOUT = toLong(System.getProperty(KEY_TIMEOUT, "5000"), 5000);
```

so please, just rever this change, which is only confusing and completely equivalent

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java [hunk 6]
```diff
@@ -64,4 +69,21 @@ public SWTBotMenu findMenu(final SWTBotMenu parentMenu,
             return new SWTBotMenu(menuItem);
         }
     }
+
+    public SWTBotMenu findWorkbenchMenu(final SWTWorkbenchBot bot, final String menuString) {
```
Comment (LorenzoBettini) [inline tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/SWTBotMenuHelper.java:73]: It might be the case to write a well-documenting Javadoc here, in particular, the problems it fixes

**Review by LorenzoBettini — CHANGES_REQUESTED** [review]

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (echebbi) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java]: > I think that we should use the same timeout consistently and make it configurable from the outside as we do right now

@LorenzoBettini I'm not sure I understood you, what do you mean exactly by "make it configurable from the outside"? As you said above the timeout is already configurable through the `org.eclipse.swtbot.search.timeout` property.

I agree that we should keep the timeout as consistent as possible thoughout the tests and avoiding changing it programmatically. Ideally I'd actually like to have a dedicated property that defines the timeout before we stop to wait for PIT so that we can change it without impacting the timeout for UI elements; something like `org.pitest.pitclipse.tests.pit.timeout`, but we'll see that it in another PR.

@JKutscha it looks like you reverted this change in your latest commit and the build seems to succeed so I guess this conversation can be closed? If you do need to increase the timeout, we set it [here](https://github.com/pitest/pitclipse/blob/157167faca6044b311f4a0606d0184770019ba45/tests/org.pitest.pitclipse.ui.tests/pom.xml#L89) for the Maven build:

https://github.com/pitest/pitclipse/blob/157167faca6044b311f4a0606d0184770019ba45/tests/org.pitest.pitclipse.ui.tests/pom.xml#L89

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/PitResultNotifier.java (hunk 5 above)
Comment (JKutscha) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/swtbot/PitResultNotifier.java:141]: I did not know that.

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (LorenzoBettini) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java]: > > I think that we should use the same timeout consistently and make it configurable from the outside as we do right now
> 
> @LorenzoBettini I'm not sure I understood you, what do you mean exactly by "make it configurable from the outside"? As you said above the timeout is already configurable through the `org.eclipse.swtbot.search.timeout` property.

Sorry for the misunderstanding: I meant "keep it" as it is right now :)

> 
> I agree that we should keep the timeout as consistent as possible thoughout the tests and avoiding changing it programmatically. Ideally I'd actually like to have a dedicated property that defines the timeout before we stop to wait for PIT so that we can change it without impacting the timeout for UI elements; something like `org.pitest.pitclipse.tests.pit.timeout`, but we'll see that it in another PR.

Yes, that should be trivial to implement. If you have time please open an issue or I'll do that ASAP

**Comment by LorenzoBettini** [comment]
Thank you @JKutscha !

File: tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (echebbi) [reply tests/org.pitest.pitclipse.ui.tests/src/org/pitest/pitclipse/ui/tests/PitclipsePitMutationsViewTest.java]: > If you have time please open an issue or I'll do that ASAP

Done, see #178
